[New Feature] BIP 352 Silent Payments Address Support

notTanveer opened the pull request:
## Description

Implemented support for BIP-352 Silent Payment address generation and updated the Silent Payments implementation to use the latest embit version.

This PR builds on top of Keith’s awesome work here: https://github.com/kdmukai/seedsigner/tree/bip352_silent_payments_basic_support

## Key changes
- Added Silent Payment address generation
- Updated to the latest embit version with Silent Payments support

## Refactor
- Refactored the xpub details screen — now extendable to show scan and spend key details as well
- Removed the dedicated BIP-85 numeric keyboard screen by directly instantiating the keyboard where needed

I included the refactor in this PR since the changes were minimal and closely related. For easier review, I recommend starting with the first commit, which focuses solely on Silent Payment address generation.


## Screenshots
<p float="left">
<img width="240" height="240" alt="image" src="https://github.com/user-attachments/assets/b663f908-b0f8-421d-a27a-12131dc76083" />
<img width="240" height="240" alt="image" src="https://github.com/user-attachments/assets/434c1dc7-0e4a-41ad-846b-83b3d3489e1d" />
<img width="240" height="240" alt="image" src="https://github.com/user-attachments/assets/c43b5d26-7357-4ffb-a4a3-563107db281b" />
<img width="240" height="240" alt="image" src="https://github.com/user-attachments/assets/b51b21d3-ef88-4ebe-aa5f-97807f29b88b" />
<img width="240" height="240" alt="image" src="https://github.com/user-attachments/assets/f92cff84-91b1-4ca4-b515-ae712aa66a86" />

</p>

This pull request is categorized as a:

- [x] New feature
- [ ] Bug fix
- [x] Code refactor
- [ ] Documentation
- [ ] Other

## Checklist

- [x] I’ve run `pytest` and made sure all unit tests pass before sumbitting the PR

If you modified or added functionality/workflow, did you add new unit tests?

- [ ] No, I’m a fool
- [x] Yes
- [ ] N/A

I have tested this PR on the following platforms/os:

- [x] Raspberry Pi OS [Manual Build](https://github.com/SeedSigner/seedsigner/blob/dev/docs/manual_installation.md)
- [ ] [SeedSigner OS](https://github.com/SeedSigner/seedsigner-os) on a Pi0/Pi0W board
- [ ] Other

notTanveer committed:
feat: sp address support

notTanveer committed:
feat: sp address support

* generate SP address (with labels)
* export SP descriptor
* added flow test
* refactored and added baseScreen for export and number entry

alvroble commented:
Thanks a lot @notTanveer for your work!! I haven't dived yet in the code but tested this PR on Raspberry Pi OS Manual Build:

- pytest passes fine
- It does not work when using a label. I got the following error when interacting with the keyboard:
    `UnhandledExceptionView({'error': ['UnboundLocalError', 'seed_screens.py, 918, in _run', " local variable 'keyboard_swap' referenced before assignment"]})`
- It works fine without a label.

Also, for "Export spend pubkey" and "Export scan privkey", I think it would improve the UX to first show the key before proceeding with the QR, similar to how the SP address is displayed first. Additionally, maybe it makes sense to show a `WarningScreen` before displaying the keys, depending on what others think.

notTanveer commented:
> * It does not work when using a label. I got the following error when interacting with the keyboard:
>   `UnhandledExceptionView({'error': ['UnboundLocalError', 'seed_screens.py, 918, in _run', " local variable 'keyboard_swap' referenced before assignment"]})`

Ah, I’d already fixed that error but forgot to push, should be good now!
just needed to assign keyboard_swap at the start of each loop to avoid the `UnboundLocalError`.
Appreciate you pointing it out!

kdmukai commented on src/seedsigner/helpers/embit_utils.py:
```diff
@@ -200,3 +201,40 @@ def sign_message(seed_bytes: bytes, derivation: str, msg: bytes, compressed: boo
+
+
+
+def encode_labeled_silent_payment_address(b_scan: ec.PrivateKey, B_spend: ec.PublicKey, label, network: str = "main", version: int = 0) -> str:
```
Type hint on `label` would be nice.

```python
    ..., label: int | str | bytes, ...
```

The convention seems to be to have spaces between each "|" which I hate, but, alas, a lot of these conventions are annoying (e.g. I would prefer to have no whitespace around `=` when setting default vals for input args).

kdmukai replied:
That being said, I'm not sure having a separate function for labels is necessary. It's probably cleaner to just have an optional `label` arg in `encode_silent_payment_address` that defaults to `None`.

notTanveer replied:
> ```python
> ..., label: int | str | bytes, ...
> ```

since KeyboardScreen(used in LabelEntryView) always gives back a string, wondering if this should just be str across the board?

alvroble commented:
Tested on Raspberry Pi OS manual build. Really neat UI. Functionally LGTM.

I'd suggest to add screenshots for all the new screens, including the warning screen before showing the scan key and both spend and public key screenshots.

kdmukai commented:
I only briefly scanned through the `BaseKeyboardEntryScreen` refactor but it triggered a few thoughts:

As far as I know, BIP-352 says basically nothing about what can or cannot be in a label. It lists incrementing integers as the most foolproof, trivially recoverable approach (i.e. every SP wallet can just automatically scan out to label `100000` to see if any of those labels had received a payment; apparently that's not as resource intensive as it sounds(?)). But Josie said that people can use labels however they like (any arbitrary text string), but then future recoverability is completely up to them (e.g. if I use a complex passphrase as a label, I better have that recorded somewhere or I'll never be able to find that label -- nor their payments --- ever again).

So it seems overly restrictive to assume that only upper/lower/digit chars are valid. Josie's first example of a possible non-incrementing integer use case was a user account number. But what if the account numbers are specified like: `f347-99337-b83x1`? Or even street address + unit number + last name to track tenants: `123 Main, #34: Simmons`.

IF we allow arbitrary text labels, it feels like we should allow common symbols as well.

Which would mean the keyboard would be the same as the existing bip39 passphrase keyboard. Which would mean there's no need to refactor (other than to rename that keyboard class to be more general).

THAT ALL BEING SAID... my sense so far is that the best practice is to assume the incrementing integer approach. There are so few wallet implementations that there's not enough data points to even suggest what an emerging consensus might be. But the incrementing integer approach was explained to me as the "don't shoot yourself in the foot" approach (because it's trivially easy for a wallet to recover those kinds of labels).

This all leads me to think that maybe this is a better approach:
* DISABLE label support entirely unless explicitly enabled; it's a niche add-on to an already niche feature. The vast majority of early adopters will just use the base SP address without wading into the more foot-gun-y waters of labels. Better to hide it away from the UX entirely unless the user specifically goes looking to enable it in Settings. Maybe to avoid having two separate settings to configure, one setting could have three options:
  * Enable SP (the option we hope that most people select)
  * Enable SP + labels (the more advanced option)
  * Disable SP (default)
* And limit our initial label implementation to JUST the integer use case.
* Leave a `TODO` note that we should revisit broadening the label support to any arbitrary char string at a later point.

And then if some of the keyboard cleanup offers real improvements (DRY, readability, flexibility) without going too far into unnecessary premature optimization, then consider pulling those changes out into their own isolated refactor PR. As it is, those changes make this PR much harder to review and a bigger "ask" to get merged.

notTanveer commented:
I agree completely, if users don’t store or record their labels, there’s a high risk they’ll forget them over time. In fact, I initially wanted to add a warning screen when labels are chosen. I started out considering only numeric labels, but ultimately decided to allow any choice. To steer users toward numeric labels, though, I set the keyboard default to digits, and that required refactoring the keyboard screen to accommodate only numbers and letters.

For symbols, I followed the Bitcoin design convention, which only includes alphanumeric characters, so that’s the approach I took. The BIP itself doesn’t explicitly limit label formats, but it does emphasize integers: “Let Bm = Bspend + hash(bscan || m)·G where m is an incrementable integer starting from 1.” It made sense to stick with integers, which again led to needing a keyboard refactor to avoid redundant code.

All that said, I think I’ll move forward with the approach you described. As a first step, I’ll keep the initial pull request small—just basic Silent Payment address generation.
Should I add label support in a subsequent PR?

kdmukai commented:
I still think integer-only labels have a place here (though, again, users should be encouraged to leave labels disabled).

But I'd like to hear what others think.

kdmukai commented on src/seedsigner/views/seed_views.py:
```diff
@@ -1224,6 +1232,282 @@ def run(self):
+        if self.settings.get_value(SettingsConstants.SETTING__DIRE_WARNINGS) != SettingsConstants.OPTION__DISABLED:
+            selected_warning_option = self.run_screen(
+                DireWarningScreen,
+                title=_("DANGER!"),
```
Style guideline: All TopNav titles should use "title case" (capitalize the first letter of each word).

kdmukai commented on src/seedsigner/views/seed_views.py:
```diff
@@ -1224,6 +1232,282 @@ def run(self):
+                DireWarningScreen,
+                title=_("DANGER!"),
+                status_headline=_("Exposing Private Key!"),
+                text=_("Your scan private key can be used to see ALL your silent payments. Keep it secret and secure!"),
```
You can't keep the scanning pk "secret" when you're expected to load it into an internet-connected SP wallet.

Instead, the scanning key acts more like an xpub (possible privacy leak but it can't spend your funds) and so:
* the warning should use similar language as the xpub export warning
* it shouldn't be a `DireWarningScreen` (red, danger) but rather the yellow warning level.

kdmukai commented on src/seedsigner/views/seed_views.py:
```diff
@@ -1224,6 +1232,282 @@ def run(self):
+                title=_("DANGER!"),
+                status_headline=_("Exposing Private Key!"),
+                text=_("Your scan private key can be used to see ALL your silent payments. Keep it secret and secure!"),
+                button_data=[ButtonOption(_("I Understand, Continue"))]
```
Style guideline: Should be consistent with other warning screens' next/continue text.

kdmukai commented on src/seedsigner/gui/screens/seed_screens.py:
```diff
@@ -570,6 +570,23 @@ def __post_init__(self):
+    payment_address: str = None
+
+    def __post_init__(self):
+        self.title = "Payment Address"
```
I would continue to use the "Silent Payment" terminology wherever possible in this flow. Therefore: "Silent Payment Address".

The long TopNav title will autoscroll.

notTanveer commented:
updated the labels to support only integers for now. also did a small refactor on the details and labels screens to clean up some redundancy.

now that i think about it — should we consider setting a limit on the label value? like 100K or maybe u32?

EDIT: now that labels have a separate enable button, should i get rid of the 'Add Label' screen?

kdmukai commented on src/seedsigner/gui/screens/seed_screens.py:
```diff
@@ -21,6 +21,91 @@
 
 
+@dataclass
+class _NumericInputScreen(KeyboardScreen):
```
I'm not sure if there's a python naming convention for starting a class name with an underscore.

Regardless, our style convention is to simply use `Base` to convey a base class that isn't expected to be instantiated. Usually the class includes functions that need to be overridden which throw an Exception in the base class (e.g. "Must but implemented in a child class").

Also if it derives from `KeyboardScreen`, the class name should reflect that:

```suggestion
class BaseNumericKeyboardScreen(KeyboardScreen):
```

kdmukai replied:
That being said, it feels like this Screen could just be instantiated without any child classes. Just some custom values passed into the constructor.

Other thoughts:
* Is there even enough here to justify being its own class? I lean to maybe just barely yes, but not completely sure.
* If this IS generally useful, why is it in `seed_screens.py` and not in a more general place like screens.py?

kdmukai commented on src/seedsigner/gui/screens/seed_screens.py:
```diff
@@ -21,6 +21,91 @@
+        self.cols = 5
+        self.keys_charset = "0123456789"
+        self.show_save_button = True
+        # Title should be set by subclasses before calling super().__post_init__()
```
If this is a requirement, I would check the value and throw an Exception if it isn't set.

kdmukai commented on src/seedsigner/gui/screens/seed_screens.py:
```diff
@@ -21,6 +21,91 @@
+    def __post_init__(self):
+        # Ensure title and button_data are set by subclass, as ButtonListScreen expects them.
+        if not hasattr(self, 'title') or not self.title:
+            raise NotImplementedError(f"{self.__class__.__name__} must set self.title in __post_init__ before super()")
```
`NotImplementedError` is meant to be user-facing to alert users of features that aren't yet supported.

In this case you can just raise an Exception since this is meant to alert a developer during coding.

kdmukai commented on src/seedsigner/gui/screens/seed_screens.py:
```diff
@@ -556,17 +641,82 @@ def __post_init__(self):
     def __post_init__(self):
         self.title = _("BIP-85 Index")
-        self.user_input = ""
+        super().__post_init__()
```
No need for this to be its own Screen. Just have the View instantiate the numeric keyboard screen directly.

kdmukai commented on src/seedsigner/views/seed_views.py:
```diff
@@ -589,6 +590,10 @@ def run(self):
             button_data.append(self.BIP85_CHILD_SEED)
 
+        if self.settings.get_value(SettingsConstants.SETTING__BIP352_SILENT_PAYMENTS) in [SettingsConstants.BIP352__ENABLED, SettingsConstants.BIP352__ENABLED_WITH_LABELS]:
+            if self.seed.get_fingerprint(self.settings.get_value(SettingsConstants.SETTING__NETWORK)) != "":
```
I don't understand this test. When would the seed's fingerprint be empty?

kdmukai commented on src/seedsigner/views/seed_views.py:
```diff
@@ -1224,6 +1232,265 @@ def run(self):
+        if label:
+            return Destination(SeedBIP352GeneratePaymentAddressView, view_args={"seed_num": self.seed_num, "label": label})
+        else:
+            return Destination(SeedBIP352GeneratePaymentAddressView, view_args={"seed_num": self.seed_num})
```
Could remove the if/else logic and just set `"label": label if label else None`.

But taking a step back, if the user enters no label at all, I think it should be handled like https://github.com/SeedSigner/seedsigner/pull/790

notTanveer replied:
> But taking a step back, if the user enters no label at all, I think it should be handled like https://github.com/SeedSigner/seedsigner/pull/790

bit of an overkill for labels 
EDIT: btw, KEY3 is disabled until there's an input from the user.

kdmukai commented on src/seedsigner/models/seed.py:
```diff
@@ -164,6 +165,42 @@ def get_bip85_child_mnemonic(self, bip85_index: int, bip85_num_words: int, netwo
+    def generate_bip352_silent_payment_labeled_address(self, label: str, network: str = SettingsConstants.MAINNET):
+        scanning_pk = self.derive_bip352_scanning_key(network=network)
+        signing_pubkey = self.derive_bip352_signing_key(network=network).get_public_key()
+        return embit_utils.encode_labeled_silent_payment_address(scanning_pk, signing_pubkey, label, network=SettingsConstants.map_network_to_embit(network))
```
I still think these should be combined and have `label` defaulting to `None`.

And similarly `embit_utils` should only have one form where a `label` arg is optional/default `None`.

kdmukai commented on src/seedsigner/models/seed.py:
```diff
@@ -164,6 +165,42 @@ def get_bip85_child_mnemonic(self, bip85_index: int, bip85_num_words: int, netwo
+
+
+    def derive_bip352_signing_key(self, account: int = 0, network: str = SettingsConstants.MAINNET):
+        return self._derive_bip352_key(is_scanning_key=False, account=account, network=network)
```
These should include a return type hint.

kdmukai commented on src/seedsigner/views/seed_views.py:
```diff
@@ -1224,6 +1232,265 @@ def run(self):
+
+
+    def get_next_destination(self):
+        return Destination(SeedBIP352SilentPaymentsOptionsView, view_args={"seed_num": self.seed_num}, clear_history=True)
```
For clarity, I would put this base class immediately above its child classes.

kdmukai commented on src/seedsigner/gui/screens/seed_screens.py:
```diff
@@ -21,6 +21,91 @@
+        current_y = self.top_nav.height + GUIConstants.COMPONENT_PADDING
+
+        # Set up the fingerprint and passphrase displays
+        if self.fingerprint:
```
Is this Screen ever used without a fingerprint?

kdmukai commented on src/seedsigner/gui/screens/seed_screens.py:
```diff
@@ -21,6 +21,91 @@
+            self.components.append(self.fingerprint_line)
+            current_y = self.fingerprint_line.screen_y + self.fingerprint_line.height + int(1.5 * GUIConstants.COMPONENT_PADDING)
+
+        if self.derivation_path:
```
Same question for derivation path.

kdmukai commented on src/seedsigner/gui/screens/seed_screens.py:
```diff
@@ -556,17 +641,82 @@ def __post_init__(self):
+@dataclass
+class SeedBIP352ExportScanningPrivkeyDetailsScreen(_BaseKeyDetailsScreen):
+    scanning_privkey: str = "xprv..."
+    DEFAULT_DERIVATION_PATH: str = "m/352'/0'/0'/1'/0" 
```
Screens should be as dumb as possible and basically know nothing about any bitcoin-y details. Give them data to display and they format it for display, that's it (ideally).

This Screen will always get its derivation path from its calling View. No need to provide a default here.

kdmukai replied:
And actually, with that bit of logic removed, there's no reason to define this child class (nor its siblings). The View can directly instantiate the Key Details Screen by just passing in params.

kdmukai commented on src/seedsigner/gui/screens/seed_screens.py:
```diff
@@ -556,17 +641,82 @@ def __post_init__(self):
+@dataclass
+class SeedBIP352ExportSigningPubkeyDetailsScreen(_BaseKeyDetailsScreen):
+    signing_pubkey: str = "xpub..."
+    DEFAULT_DERIVATION_PATH: str = "m/352'/0'/0'/0'/0"
```
Same issue.

kdmukai commented on src/seedsigner/gui/screens/seed_screens.py:
```diff
@@ -604,62 +754,23 @@ def __post_init__(self):
-    derivation_path: str = "m/84'/0'/0'"
+class SeedExportXpubDetailsScreen(_BaseKeyDetailsScreen):
     xpub: str = "zpub6r..."
+    DEFAULT_DERIVATION_PATH: str = "m/84'/0'/0'" 
```
Same as above.

kdmukai commented:
Bug: If I enter a label as `0000`, I get a different SP result than if I enter `0`.

Same for '01' vs '1', etc.

kdmukai commented:
Remind me how it all works: Adding a label changes the shareable SP address obv. But do labels affect the scan or spend key at all?

Just trying to refresh my memory and think through the scan privkey and spend pubkey export tasks.

notTanveer commented:
> Bug: If I enter a label as `0000`, I get a different SP result than if I enter `0`.
> 
> Same for '01' vs '1', etc.

the KeyboardScreen component accepts string input for labels, treating 01 and 1 as distinct string values rather than equivalent numeric inputs.

I'd prefer not to modify KeyboardScreen. what's the best approach to handle this @kdmukai ?

EDIT:

ig this is possible, we could just cast the label to an int inside the method `encode_silent_payment_address`. (feels kinda hacky, tbh)

notTanveer commented:
> Remind me how it all works: Adding a label changes the shareable SP address obv. But do labels affect the scan or spend key at all?
> 
> Just trying to refresh my memory and think through the scan privkey and spend pubkey export tasks.

scan and spend key are derived from the master seed, and labels have no role in modifying those, labels only add extra data at the end while generating the sp address.

kdmukai commented on src/seedsigner/gui/screens/screen.py:
```diff
@@ -1322,3 +1323,90 @@ class MainMenuScreen(LargeButtonScreen):
+
+
+@dataclass
+class BaseKeyDetailsScreen(WarningEdgesMixin, ButtonListScreen):
```
The name "KeyDetails" wasn't immediately clear to me; as a former English teacher, we often talk about "key details" in reading assignments, etc. So that term just doesn't trigger the intended connection to pubkey/privkey.

I don't know what a better name would be and this is possibly just a "me" issue.

notTanveer replied:
how about **BaseExportKeyDetailsScreen** ?

kdmukai commented on src/seedsigner/gui/screens/screen.py:
```diff
@@ -1322,3 +1323,90 @@ class MainMenuScreen(LargeButtonScreen):
+@dataclass
+class BaseKeyDetailsScreen(WarningEdgesMixin, ButtonListScreen):
+    """
+    Standardized base class for displaying key details.
```
Here I would elaborate with "pubkey or privkey details".

kdmukai commented on src/seedsigner/gui/screens/screen.py:
```diff
@@ -1322,3 +1323,90 @@ class MainMenuScreen(LargeButtonScreen):
+    """
+    Standardized base class for displaying key details.
+    
+    Args:
+        title: Screen title
+        key_label: Label for the key (e.g., "Xpub", "Private Key")
+        key_value: The actual key value to display
+        derivation_path: BIP32 derivation path
+        fingerprint: BIP32 master fingerprint
+        has_passphrase: Whether seed has passphrase (for warning styling)
+        button_data: List of ButtonOption objects
+        icon_name: FontAwesome icon for the key line (defaults to 'x')
```
There's no harm here, but we don't document every Screen like this. It's more important to just document the unique args.

kdmukai commented on src/seedsigner/gui/screens/screen.py:
```diff
@@ -1322,3 +1323,90 @@ class MainMenuScreen(LargeButtonScreen):
+        if not self.key_value:
+            raise Exception("Implement key_value in subclass")
+        if not self.button_data:
+            raise Exception("Implement button_data in subclass")
```
Bulletproofing never hurts, but args like `title` and `button_data` are ubiquitous in the code so are unnecessary call outs here. I wouldn't worry about any of these, tbh.

kdmukai commented on src/seedsigner/helpers/embit_utils.py:
```diff
@@ -200,3 +201,31 @@ def sign_message(seed_bytes: bytes, derivation: str, msg: bytes, compressed: boo
+
+
+# Basic BIP-352 Silent Payments support
+def encode_silent_payment_address(scan_privkey: ec.PrivateKey, spend_pubkey: ec.PublicKey, label: int | str | bytes | None = None, network: str = "main", version: int = 0) -> str:
```
We always differentiate our network setting vs `embit_network` (see other methods in this file).

kdmukai commented on src/seedsigner/helpers/embit_utils.py:
```diff
@@ -200,3 +201,31 @@ def sign_message(seed_bytes: bytes, derivation: str, msg: bytes, compressed: boo
+        * optional label for labeled addresses
+
+    if scan_key is a PrivateKey and label is provided, generates a labeled address.
+    """
```
Maybe add a `TODO` note that we'd like to eventually see this functionality moved into `embit`.

kdmukai commented on src/seedsigner/models/seed.py:
```diff
@@ -164,6 +165,38 @@ def get_bip85_child_mnemonic(self, bip85_index: int, bip85_num_words: int, netwo
+
+    def derive_bip352_scan_privkey(self, network: str = SettingsConstants.MAINNET) -> bip32.HDKey:
+        return self._derive_bip352_key(is_scan_key=True, network=network)
+
```
Our opinionated style is to use two blank lines between member functions.

kdmukai commented on src/seedsigner/models/seed.py:
```diff
@@ -164,6 +165,38 @@ def get_bip85_child_mnemonic(self, bip85_index: int, bip85_num_words: int, netwo
         return bip85.derive_mnemonic(root, bip85_num_words, bip85_index)
         
 
+    # ----------------- BIP-352 Silent Payments support -----------------
```
I think these methods are fine here for now. But in general we have pretty fuzzy distinctions between what operations are in `Seed` vs `embit_utils` and ultimately what we hope to be in `embit` itself.

These feel like they should eventually be in `embit`.

kdmukai commented on src/seedsigner/models/settings_definition.py:
```diff
@@ -399,6 +400,17 @@ def map_network_to_embit(cls, network) -> str:
 
+    # BIP-352 Silent Payments options
+    BIP352__DISABLED = "D"
+    BIP352__ENABLED = "E"
```
I don't think you should redefine new enabled/disabled constants. We can just re-use the existing ones.

kdmukai commented on src/seedsigner/models/settings_definition.py:
```diff
@@ -699,6 +711,15 @@ class SettingsDefinition:
                       visibility=SettingsConstants.VISIBILITY__ADVANCED,
                       default_value=SettingsConstants.OPTION__ENABLED),
 
+        SettingsEntry(category=SettingsConstants.CATEGORY__FEATURES,
```
I would move this to be just below the BIP-85 settings. The settings down here are all about turning on and off UI displays.

Eventually it probably makes sense to list all the "BIP-*" settings after each other, but that should be left to a separate PR.

kdmukai replied:
(or eventually drop the "BIP" references altogether since they're not very user friendly)

kdmukai commented on src/seedsigner/views/seed_views.py:
```diff
@@ -1172,8 +1184,14 @@ def __init__(self, seed_num: int, num_words: int):
     def run(self):
-        # TODO: Change this later to use the generic Screen input keyboard
-        ret = seed_screens.SeedBIP85SelectChildIndexScreen().display()
+        ret = self.run_screen(
+            seed_screens.KeyboardScreen,
+            title=_("BIP-85 Child Index"),
+            rows=3,
+            cols=5,
+            keys_charset="0123456789",
+            show_save_button=True,
+        )
```
Is this a leftover edit from an earlier version? If so, I would revert this to keep this PR as focused as possible. Same for the removal of the `SeedBIP85SelectChildIndexScreen`.

notTanveer replied:
IIRC, you mentioned not to use the custom `NumericScreen` and instead just instantiate the **`KeyboardScreen` directly** in the **view** with all the custom configurations.

I initially created a custom NumericInputScreen for numeric inputs when i needed a numeric screen for labels. however, since you pointed out that we can easily instantiate this directly in the view, I refactored it accordingly.

the only other place using a custom input screen was SeedBIP85SelectChildIndexScreen, so I applied the same approach there as well.

This is a relatively small change, so I believe it makes sense to include it in this PR itself.

Chaitanya-Keyal replied:
Isn't `SeedBIP85SelectChildIndexScreen` already using the generic `KeyboardScreen`?

https://github.com/SeedSigner/seedsigner/blob/49dea74eb32c3cddbfcc039224af7e9af2d6eb0d/src/seedsigner/gui/screens/seed_screens.py#L558C1-L570C32

`SeedExportXpubCustomDerivationScreen` is also implemented in a similar manner, simply extending `KeyboardScreen`.

I think the `TODO` comment here was leftover from the initial BIP-85 PR. See: https://github.com/SeedSigner/seedsigner/pull/785#discussion_r2328353516

notTanveer replied:
> Isn't `SeedBIP85SelectChildIndexScreen` already using the generic `KeyboardScreen`?

it does, but why create a separate screen class when you can instantiate it directly in the view? (reduces code debt)

> `SeedExportXpubCustomDerivationScreen` is also implemented in a similar manner, simply extending `KeyboardScreen`

this can be definitely done, thanks for pointing this out.

Chaitanya-Keyal replied:
Fair I guess, but yeah it would probably make sense to do all of this in a separate PR.

kdmukai commented on src/seedsigner/views/seed_views.py:
```diff
@@ -1226,6 +1244,249 @@ def run(self):
+    BIP-352: Silent Payments
+****************************************************************************"""
+class SeedBIP352SilentPaymentsOptionsView(View):
+    GENERATE_SP_ADDRESS = ButtonOption("Generate SP address")
```
This needs a `TRANSLATOR_NOTE` to explain what "SP" is.

kdmukai commented on src/seedsigner/views/seed_views.py:
```diff
@@ -1226,6 +1244,249 @@ def run(self):
+    BIP-352: Silent Payments
+****************************************************************************"""
+class SeedBIP352SilentPaymentsOptionsView(View):
+    GENERATE_SP_ADDRESS = ButtonOption("Generate SP address")
+    EXPORT_SPEND_PUBKEY = ButtonOption("Export spend pubkey")
+    EXPORT_SCAN_PRIVKEY = ButtonOption("Export scan privkey")
```
Additional `TRANSLATOR_NOTE`s for each of these would be helpful, too. A translator might know nothing about Silent Payments and will be unfamiliar with what is being listed here.

kdmukai commented on src/seedsigner/views/seed_views.py:
```diff
@@ -1226,6 +1244,249 @@ def run(self):
+
+        selected_menu_num = self.run_screen(
+            ButtonListScreen,
+            title=_("Silent Payments"),
```
Another place for a `TRANSLATOR_NOTE`, just to make sure the translator realizes this is referring to a specific technique / BIP and that their translation should match how the BIP is referred to in their language.

kdmukai commented on src/seedsigner/views/seed_views.py:
```diff
@@ -1226,6 +1244,249 @@ def run(self):
+    def run(self):
+        selected_option = self.run_screen(
+            BaseKeyDetailsScreen,
+            title=_("Spend Pubkey Details"),
+            key_label=_("Spend Public Key"),
```
More places to drop `TRANSLATOR_NOTE` comments. It is repetitive, but the translation strings are presented to the translators in seemingly random order. So we have no guarantees if this is the first time they're ever encountering the terminology while translating or if it's the 20th.

kdmukai commented on src/seedsigner/views/seed_views.py:
```diff
@@ -1226,6 +1244,249 @@ def run(self):
+                WarningScreen,
+                title=_("Caution!"),
+                status_headline=_("Privacy Leak!"),
+                text=_("Your scan private key can be used to detect all your silent payments but cannot spend funds."),
```
`TRANSLATOR_NOTE` to at least indicate this is part of the bip-352 flow.

kdmukai commented on src/seedsigner/views/seed_views.py:
```diff
@@ -1226,6 +1244,249 @@ def run(self):
+                return Destination(BackStackView)
+
+        selected_details_option = self.run_screen(
+            BaseKeyDetailsScreen,
+            title=_("Scan Private Key Details"),
+            key_label=_("Scan Private Key"),
```
`TRANSLATOR_NOTE` - this one could be misinterpreted as a command ("Scan your private key now" instead of talking about the "scan private thing" as a noun). Have to make sure the translators aren't confused about the intention here.

kdmukai commented on src/seedsigner/views/seed_views.py:
```diff
@@ -1226,6 +1244,249 @@ def run(self):
+        super().__init__()
+        self.seed_num = seed_num
+        
+    #TODO: extend support for alphabets and special characters
```
I would phrase this as something like "arbitrary (non-integer) labels"

kdmukai commented on src/seedsigner/views/view.py:
```diff
@@ -414,3 +414,23 @@ def run(self):
+        
+
+
+class BaseQRDisplayView(View):
```
Took me a while to decide if there was enough of a DRY/optimization win here, and I think there just barely is.

It's used well for the new QR displays in this PR. But I would leave a `TODO` note that other QR display Views should maybe be refactored to use this as well.

kdmukai commented:
This PR definitely needs to add a flow test to be considered complete!

I'd suggest:
* Flow with labels disabled.
* Flow with a custom label or two.

kdmukai commented on tests/test_flows_seed.py:
```diff
@@ -734,3 +734,151 @@ def expect_unsupported_derivation(load_message: Callable):
+        )
+
+
+    def test_bip352_generate_sp_address_with_labels_no_label_flow(self):
```
Super minor nit: "with labels no label" is confusing. Might be clearer in all of these test names to use "labels_enabled" / "labels_disabled" and then maybe something like "skip_label" / "specify_label"? Or "with_label" / "without_label"? You get the idea...

Also elsewhere we use double underscore to help provide extra clarity / separation in names when needed. I don't know if any flow tests use it, but I find it useful. e.g. `labels_enabled__skip_label`...

kdmukai commented on tests/test_flows_seed.py:
```diff
@@ -734,3 +734,151 @@ def expect_unsupported_derivation(load_message: Callable):
+                FlowStep(seed_views.SeedBIP352SilentPaymentsOptionsView, button_data_selection=seed_views.SeedBIP352SilentPaymentsOptionsView.EXPORT_SPEND_PUBKEY),
+                FlowStep(seed_views.SeedBIP352ExportSpendPubkeyDetailsView, screen_return_value=0),
+                FlowStep(seed_views.SeedBIP352ExportSpendPubkeyQRView, screen_return_value=0),
+                FlowStep(seed_views.SeedBIP352SilentPaymentsOptionsView),
+            ]
+        )
+
+
+    def test_bip352_export_scan_privkey_flow(self):
+        mnemonic = "blush twice taste dawn feed second opinion lazy thumb play neglect impact".split()
+        self.controller.storage.set_pending_seed(Seed(mnemonic=mnemonic))
+        self.controller.storage.finalize_pending_seed()
+
+        self.settings.set_value(SettingsConstants.SETTING__BIP352_SILENT_PAYMENTS, SettingsConstants.OPTION__ENABLED)
+
+        self.run_sequence(
+            initial_destination_view_args=dict(seed_num=0),
+            sequence=[
+                FlowStep(seed_views.SeedOptionsView, button_data_selection=seed_views.SeedOptionsView.BIP352_SILENT_PAYMENTS),
```
I probably would have just directly connected these tests into one sequence.

The strict separation here is slightly more rigorous, of course, just a tiny bit unnecessary.

kdmukai commented on tests/test_flows_seed.py:
```diff
@@ -734,3 +734,151 @@ def expect_unsupported_derivation(load_message: Callable):
+    """
+
+    def test_bip352_silent_payments_disabled_flow(self):
+        mnemonic = "blush twice taste dawn feed second opinion lazy thumb play neglect impact".split()
+        self.controller.storage.set_pending_seed(Seed(mnemonic=mnemonic))
+        self.controller.storage.finalize_pending_seed()
```
I would leverage the `pytest` harness to do this in something like `setup_class` (I can't recall the exact method name) where it's done once for the entire class before any of its child tests are run and is then present for all tests.

There's another method that's run before each individual test. I can't recall which scope would be required for this sort of pre-test config.

kdmukai commented:
Can you regenerate the screenshots and update them in the PR description? I know at least the "DANGER" screen has changed.

kdmukai commented on src/seedsigner/views/seed_views.py:
```diff
@@ -1229,6 +1247,269 @@ def run(self):
+****************************************************************************"""
+class SeedBIP352SilentPaymentsOptionsView(View):
+    # TRANSLATOR_NOTE: These options are related to BIP-352 Silent Payments(SP).
+    # the first option allows the user to generate a static reusable SP address.
+    # next two options help in exporting the spend public key and scan private key, 
+    # which is important for detecting incoming payments.
```
The translators don't necessarily experience the translation strings in this sort of linear fashion.

Each TRANSLATOR_NOTE is only displayed for the translation string that immediately follows it; the subsequent entries won't have any context hints. It's best to treat each translation string as being totally devoid of any surrounding / sequential context.

kdmukai commented on src/seedsigner/views/seed_views.py:
```diff
@@ -1229,6 +1247,269 @@ def run(self):
+    # the first option allows the user to generate a static reusable SP address.
+    # next two options help in exporting the spend public key and scan private key, 
+    # which is important for detecting incoming payments.
+    GENERATE_SP_ADDRESS = ButtonOption("Generate SP address")
```
What if we just added an option here for "SP address + label" (or something like that) in order to avoid the label Yes/No screen entirely?

kdmukai replied:
Also worth thinking about if "Generate SP address" could offer a tiny bit more guidance: "Shareable SP address"?

Or just: "Silent payment address"? Or "Export SP address" to match our xpub approach?

Basically, "Generate" is taking up a lot space but isn't really useful.

notTanveer commented:
I believe i've addressed all the feedback.

## Changes:

- Removed `**LabelPromptScreen**` and added dedicated SP labeled address button
- Added Flow Tests with pytest `setup_method` to instantiate test seed and reduce redundancies
- Enforced consistent integer-only SP address label generation
- Updated translator notes
- Renamed test flow methods following better naming conventions

## Questions:

- Given our focus on UX and foolproof labels, should we add integer limits for now? (BIP doesn't specify any constraints)
- currently type-casting string input from label screen to integers since **KeyboardScreen** only accepts strings, but our labels support int/str/bytes types. Should we standardize on integers / strings throughout instead of this conversion? (i don't want to modify KeyboardScreen )
- should I update the messages.pot file?

newtonick commented:
cACK [1d984f8](https://github.com/SeedSigner/seedsigner/pull/769/commits/1d984f8b6c6d044e5c789f7bd64ed28d9524aa96)

I haven't had the time to dive into this code, but did run the PR and did some functionality testing. Looks really good. Looking forward to getting SP functionality merged into SeedSigner. Thanks for this work!

jvgelder commented on tests/test_embit_utils.py:
```diff
@@ -426,3 +429,75 @@ def test_parse_derivation_path():
+    LABEL_TEST_VECTORS = {
+    "spend_priv_key": "9d6ad855ce3417ef84e836892e5a56392bfba05fa5d97ccea30e266f540e08b3",
+    "scan_priv_key": "0f694e068028a717f8af6b9411f9a133dd3565258714cc226594b34db90c1f2c",
+    "labels": [
```
labels should be incremental integers:
see: m is an incrementable integer starting from 1 see https://github.com/bitcoin/bips/blob/ae9cd401dfe6d82c4759975fe5c6ecc31376d253/bip-0352.mediawiki?plain=1#L117

kdmukai replied:
I think you might be misunderstanding the point of this test.

Regardless, this test vector is taken directly from the BIP itself: https://github.com/bitcoin/bips/blob/master/bip-0352/send_and_receive_test_vectors.json#L1446-L1450

jvgelder replied:
Good point, for the reasons you mentioned in [th](https://github.com/SeedSigner/seedsigner/pull/769#discussion_r2556419879) might be wise if its more explicitly stated in the bip and the test vectors also adhere to it ;) This also make it easy to store as you can just record maxlabel.

jvgelder commented on tests/test_embit_utils.py:
```diff
@@ -426,3 +429,75 @@ def test_parse_derivation_path():
+        sp_address = embit_utils.encode_silent_payment_address(scan_priv_key, spend_priv_key.get_public_key(), label)
+        assert sp_address == address
+
+    # Label may also be a string, but the bip does not provide any test vectors
```
it needs to be an integer

kdmukai replied:
I don't know if Josie has changed his opinion more recently, but during extensive discussions with him he has said that a label could be anything. His example was a large exchange could use an alphanumeric user ID as a label.

Obviously that comes with risks; it is then up to the recipient to keep a record of all such labels since you lose the simple guarantees of an incrementing integer.

That being said, since that use case isn't described in the BIP, I'd be fine dropping all mention of any alternate use case (i.e. only supporting incrementing integer labels). I think that's where discussions in this PR had landed already.

In which case @notTanveer can simply remove this test case and the following one.

notTanveer replied:
yep, i guess that makes sense. 

also, i believe it's good to update the `encode_silent_payment_address()` function to deal with unnecessary overhead.

notTanveer replied:
updated. refer to the changes made [here](https://github.com/SeedSigner/seedsigner/compare/a4ae27d79e238496e7bac94f341026b26fdd4b22..738d5ab9c5e44f876164e899c853777a3fd6540e)

jvgelder commented on tests/test_embit_utils.py:
```diff
@@ -426,3 +429,75 @@ def test_parse_derivation_path():
+    # Label may also be a string, but the bip does not provide any test vectors
+    embit_utils.encode_silent_payment_address(scan_priv_key, spend_priv_key.get_public_key(), label="tenant 6102")
+
+    # Label may also be passed in as bytes
```
same here

macgyver13 commented:
Linking [sp() output descriptor](https://github.com/bitcoin/bips/pull/2047) bip proposal for visibility

Great to see the progress made on Label UI/UX

earthdiver commented:
> > Linking [sp() output descriptor](https://github.com/bitcoin/bips/pull/2047) bip proposal for visibility
> 
> now that [BIP-392](https://github.com/bitcoin/bips/blob/master/bip-0392.mediawiki) is in, the question is: do we still keep the existing export of scan/spend private keys, or move to the descriptor-based approach (although, still in draft and no test vectors) ?

In Sparrow Wallet v25.x, receiving Silent Payments is now supported, and configuring a watch-only wallet for this requires specifying spscan.

I would appreciate it if you could add support for displaying spscan.

notTanveer commented:
> I would appreciate it if you could add support for displaying spscan.

that’s the plan, especially since we finally have a coordinator to properly test against.

after thinking it through, address generation and export support only really make sense once receive support exists and you’re actually able to spend your silent payment outputs. i think that part fits better as a separate wallet type altogether (probably single-sig SP).

so yeah, right now i’m mainly focused on send-side SP support. my embit PRs are also starting to gain momentum, and once those land, adding support for sending from a normal wallet to an SP address. and later spending received SP outputs. after that, we can revisit this PR and finally get it merged in with the agreed spscan approach.

earthdiver commented:
> > > Linking [sp() output descriptor](https://github.com/bitcoin/bips/pull/2047) bip proposal for visibility
> > 
> > 
> > now that [BIP-392](https://github.com/bitcoin/bips/blob/master/bip-0392.mediawiki) is in, the question is: do we still keep the existing export of scan/spend private keys, or move to the descriptor-based approach (although, still in draft and no test vectors) ?
> 
> In Sparrow Wallet v25.x, receiving Silent Payments is now supported, and configuring a watch-only wallet for this requires specifying spscan.
> 
> I would appreciate it if you could add support for displaying spscan.

For reference, I’m sharing test vectors I created based on Keith’s gist (https://gist.github.com/kdmukai/e6de87c90d4fedeaf866f2b10c0833a9), using my own implementation (https://github.com/earthdiver/psBitcoin/blob/main/examples/07_SilentPaymentAddresses_BIP352.ps1).

```
[
  {
    "mnemonic": "initial tilt corn easily leave weather strategy return topple gesture sad day",
    "coin_type": 1,  // testnet
    "payment_addr":  "tsp1qqfvn9pmvmz0ewpnp7w302lxqmnue2kgtpne2p38nuunun883sw36yq48ny7n2jl0nx9ljhmdnrgvpee6aufmg9wfvqfcr6c02at6r4u4xsegph7a",
    "payment_addr0": "tsp1qqfvn9pmvmz0ewpnp7w302lxqmnue2kgtpne2p38nuunun883sw36yqu9e9gaqt04w3svwy8v6plfevlar572y5hmj6unex9xdagzu5426gkhqm4q",
    "payment_addr1": "tsp1qqfvn9pmvmz0ewpnp7w302lxqmnue2kgtpne2p38nuunun883sw36yq7frtfktpq0ae6lsk2v32qmpm6tdu2p5at7w8grvcrtkn5ap0727uu2vl8e",
    "payment_addr2": "tsp1qqfvn9pmvmz0ewpnp7w302lxqmnue2kgtpne2p38nuunun883sw36yqk0pvxs9zd7gezdv6ctt69c72h2pjmmfj0hz0m0n8mfrg0zwfz9d583wcee",
    "spscan":  "tspscan1q09zrmaz09cdzs5jxm552qpv3f2gxd9vxhs0yady09jdd6aqt5e7s9fue8565hmue30u47mvc6rqwwwh0zw6ptjtqzwq7kr6h27sa09f5g6x977",
    "spspend": "tspspend1q09zrmaz09cdzs5jxm552qpv3f2gxd9vxhs0yady09jdd6aqt5e772yf8kwpa7shfhuw9esasvgn8lh7e6ufea60fpvfx9dk7m3klg6sa90au8",
  },
  {
    "mnemonic": "tongue vanish post gentle fever figure kangaroo select infant blur phrase relief",
    "coin_type": 0,  // mainnet
    "payment_addr":  "sp1qq2c4jvrju33tmm9ll0560vm0rflfxkhd8zj74pka8s53dyaztzwlqqhrkuv0ut7wjv08kdq26t4twguxdcd9m35p6z4n784wyg3efwruevxty23x",
    "payment_addr0": "sp1qq2c4jvrju33tmm9ll0560vm0rflfxkhd8zj74pka8s53dyaztzwlqq3nh20szx3lkpttrpx9u56ae6z7r7ejeyyh623vkdeupud385wzxczvd2xe",
    "payment_addr1": "sp1qq2c4jvrju33tmm9ll0560vm0rflfxkhd8zj74pka8s53dyaztzwlqq59wqj4cckjsq6a7wh9kj2t9extk0f0c93lsd4ewgzy7uwh953ngswzrweq",
    "payment_addr2": "sp1qq2c4jvrju33tmm9ll0560vm0rflfxkhd8zj74pka8s53dyaztzwlqqer5xely65uhjr4sqh88q2jajapej43c8x8yt4xxd99y2qq5thrfuer8j5u",
    "spscan":  "spscan1qnd95fpg2587jn73qg98pq8uk20y09v5c20u0e4kynsc4m2qmkrrs9cahrrlzln5nreangzkja2mj8pnwrfwudqws4vl3at3zyw2tslxtryq7pn",
    "spspend": "spspend1qnd95fpg2587jn73qg98pq8uk20y09v5c20u0e4kynsc4m2qmkrrmd9tyggwt47773rhumkklet4g2us7c3x0gul65za8fg32nansdesukdqr2",
  },
  {
    "mnemonic": "index today witness obscure ugly curtain symbol pumpkin pelican child maple struggle arctic water tiny pizza harbor below violin eight tennis frost clown hood",
    "coin_type": 1,  // testnet
    "payment_addr":  "tsp1qqvdcq76j5kul4s6t52d07ssq8l96k49jur0kytua36k9qzj4m5xyxq5j2v4hc8njddv9xtnhly7hyv2agt28fypqn29q8mw3fjjlz00vvv824hd6",
    "payment_addr0": "tsp1qqvdcq76j5kul4s6t52d07ssq8l96k49jur0kytua36k9qzj4m5xyxqnk5r88uv3459esc08x5hvkq9yy2chyz7cy4y6knp6xn42chm9p5u6afswf",
    "payment_addr1": "tsp1qqvdcq76j5kul4s6t52d07ssq8l96k49jur0kytua36k9qzj4m5xyxq6twmpszhkyxf44tjp4s6eh6dmvtckp4fh3t6zl80e0fwd47uyfzgy94nw2",
    "payment_addr2": "tsp1qqvdcq76j5kul4s6t52d07ssq8l96k49jur0kytua36k9qzj4m5xyxq6rphuf7ghxzqnz9ul2x4zshtsh0e73kgaau430au2uhc9cvnw0wqhtzpp2",
    "spscan":  "tspscan1q0z4tkwaar4ww77qgesalgzw0c40q89zh7p7hmp3qn73yrdw9jpvs9yjn9d7puunttpfjuale84erzh2z636fqgy63gp7m52v5hcnmmrrlrxnur",
    "spspend": "tspspend1q0z4tkwaar4ww77qgesalgzw0c40q89zh7p7hmp3qn73yrdw9jpvnadsvv7qqcd8ytmdtzn6r5ywvccgzpw2386spvymglmszzep0svg39qpev",
  },
  {
    "mnemonic": "fold cotton pipe robust eagle rabbit coach average orient utility minor absurd fine claim artist rabbit kingdom original lobster cruise march city vibrant resemble",
    "coin_type": 0,  // mainnet
    "payment_addr":  "sp1qq25f3laffnhpl69ytaxzz5gjnkrm2a2jr3mfz0ff6wuesg4j9e5lcqjj5qq7fy0t0wy9qvty7l7wk8vnmyxpxeq5ae0lmmzlgwnutg8945k2w7lh",
    "payment_addr0": "sp1qq25f3laffnhpl69ytaxzz5gjnkrm2a2jr3mfz0ff6wuesg4j9e5lcqjwf576ajhswcnu4qut3u0nmxf9tq5czttwmukfhk25xe00jl9yacfm9g7u",
    "payment_addr1": "sp1qq25f3laffnhpl69ytaxzz5gjnkrm2a2jr3mfz0ff6wuesg4j9e5lcql6jj65qatvpd0wcdxeya4akdl90c5p7t98f3rh6tmdyyzq0kds3cxqcurj",
    "payment_addr2": "sp1qq25f3laffnhpl69ytaxzz5gjnkrm2a2jr3mfz0ff6wuesg4j9e5lcq4k94apnmyu0f42n625vwdejmg36ca56ly8rsan600muk22sz2yqu9r8k63",
    "spscan":  "spscan1q79q4zljllyehszny72w5zfptzpxnp96esg0n2fwecgzd2v7fr6fsy54qq8jfr6mm3pgrze8hln43my7epsfkg98wtl77ch6r5lz6pedd2jcnxk",
    "spspend": "spspend1q79q4zljllyehszny72w5zfptzpxnp96esg0n2fwecgzd2v7fr6fs8y6dg3fu9jp5rhrycnuhtd555t6904x4xs7cklka8z5tk5p9xwq82pf6k",
  },
]
```

notTanveer committed:
bump: embit to feat/silent-payments branch

notTanveer committed:
fix(tests): append version 2, to PSBT OutputScope

the latest upgraded to embit includes PSBT v2 and will discard all the
v2 only fields, if the version = 2 is not set.
